core/sds-aspect-model-resolver/src/main/java/io/openmanufacturing/sds/aspectmodel/resolver/AspectModelResolver.java

```diff
@@ -130,85 +112,134 @@ public Try<VersionedModel> resolveAspectModel( final ResolutionStrategy resolver
+    * For meta model elements or other URIs, an empty model is returned. This method returns only a failure, when the used resolution
+    * strategy fails.
+    * @param urn the URN to resolve
+    * @param resolutionStrategy the resolution strategy to apply
+    * @return the model containing the defintion of the given model element
+    */
+   private Try<Model> getModelForUrn( final String urn, final ResolutionStrategy resolutionStrategy ) {
+      if ( urn.startsWith( RDF.getURI() ) || urn.startsWith( XSD.getURI() ) ) {
+         return Try.success( EMPTY_MODEL );
       }
 
-      addModelWithoutOverwritingEmptyPrefix( targetModel, loadedModel.get() );
-      return getAllUrnsInModel( loadedModel.get() )
-            .stream()
-            .filter( modelUrn -> !modelUrn.isBammUrn() )
-            .reduce( Tuple.of( Try.success( targetModel ), updatedResolvedUrns ),
-                  (( tuple, aspectModelUrn ) ->
-                        resolve( targetModel, aspectModelUrn, resolver, updatedResolvedUrns )),
-                  ( tuple, tuple2 ) -> {
-                     final Try<Model> mergedModels = tuple._1().flatMap( model ->
-                           tuple2._1().map( model2 -> addModelWithoutOverwritingEmptyPrefix( model, model2 ) ) );
-                     final Set<AspectModelUrn> mergedSets = Sets.union( tuple._2(), tuple2._2() );
-                     return Tuple.of( mergedModels, mergedSets );
-                  } );
+      try {
+         final AspectModelUrn aspectModelUrn = AspectModelUrn.fromUrn( urn );
+         if ( aspectModelUrn.getElementType() != ElementType.NONE ) {
+            return Try.success( EMPTY_MODEL );
+         }
+         return resolutionStrategy.apply( aspectModelUrn ).flatMap( model -> {
+            if ( !model.contains( model.createResource( urn ), RDF.type, (RDFNode) null ) ) {
+               return Try.failure( new ModelResolutionException( "Resolution strategy returned a model which does contain element definition for " + urn ) );
+            }
+            return Try.success( model );
+         } );
+      } catch ( final UrnSyntaxException e ) {
+         // If it's no valid Aspect Model URN but some other URI (e.g., a bamm:see value), there is nothing
+         // to resolve, so we return just an empty model
+         return Try.success( EMPTY_MODEL );
+      }
    }
 
-   private Model addModelWithoutOverwritingEmptyPrefix( final Model target, final Model modelToAdd ) {
-      if ( !target.getNsPrefixMap().containsKey( "" ) && modelToAdd.getNsPrefixMap().containsKey( "" ) ) {
-         target.setNsPrefix( "", modelToAdd.getNsPrefixURI( "" ) );
+   /**
+    * Defensively merge a model into an existing target model. This means:
+    * <ul>
+    *    <li>Prefixes are only added when they are not already present, i.e., a model won't overwrite the empty prefix of the target model</li>
+    *    <li>Statements that have a blank node (including RDF lists) as their object won't be added</li>
```

Comment:
Won't this cause the merged model to miss certain statements, i.e. the merged model will be incomplete?

Reply:
You are correct for RDF in general, but for BAMM this should not happen. However, you had me thinking and the complete distinction is not necessary any more with the new resolution method; because a model is never merged more than once any more, so we can not end up with situations such as duplicate "?s bamm:properties ?o" statements. I have removed the corresponding section and simplified the JavaDoc.

Reply:
Thanks for supporting this bug fix, Georg!